flang/lib/Lower/Allocatable.cpp

```diff
@@ -1087,6 +1088,29 @@ void Fortran::lower::associateMutableBoxWithShift(
       });
 }
 
+static bool
+isArraySectionWithoutVectorSubscript(const Fortran::lower::SomeExpr &expr) {
+  return expr.Rank() > 0 && Fortran::evaluate::IsVariable(expr) &&
+         !Fortran::evaluate::UnwrapWholeSymbolDataRef(expr) &&
+         !Fortran::evaluate::HasVectorSubscript(expr);
+}
+
+void Fortran::lower::associateMutableBoxWithShift(
+    Fortran::lower::AbstractConverter &converter, mlir::Location loc,
+    const fir::MutableBoxValue &box, const Fortran::lower::SomeExpr &source,
+    mlir::ValueRange lbounds, Fortran::lower::StatementContext &stmtCtx) {
+  auto &builder = converter.getFirOpBuilder();
+  if (Fortran::evaluate::UnwrapExpr<Fortran::evaluate::NullPointer>(source)) {
+    Fortran::lower::disassociateMutableBox(builder, loc, box);
+    return;
+  }
+  // Do not generate a temp in case rhs is an array section.
```

Comment:
Might be helpful to explain a bit more here. We don't need a temp in this case because descriptor majik...